lib/src/idg/u64.rs

```diff
@@ -0,0 +1,181 @@
+use crate::err::Error;
+use crate::kvs::{Key, Val};
+use roaring::RoaringTreemap;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+
+// Id is a unique id generated by the generator.
+pub(crate) type Id = u64;
+
+// U64 is a generator that generates unique unsigned 64-bit integer ids.
+// It can reuse freed ids by keeping track of them in a roaring bitmap.
+// This doesn't do any variable-length encoding, so it's not as space efficient as it could be.
+// It is used to generate ids for any SurrealDB objects that need aliases (e.g. namespaces, databases, tables, indexes, etc.)
+#[derive(Clone)]
+pub struct U64 {
+	state_key: Key,
+	available_ids: Option<RoaringTreemap>,
+	next_id: Id,
+	updated: bool,
+}
+
+impl U64 {
+	pub(crate) async fn new(state_key: Key, v: Option<Val>) -> Result<Self, Error> {
+		let state: State = if let Some(val) = v {
+			State::try_from_val(val)?
+		} else {
+			State::new()
+		};
+		Ok(Self {
+			state_key,
+			available_ids: state.available_ids,
+			updated: false,
+			next_id: state.next_id,
+		})
+	}
+
+	pub(crate) fn get_next_id(&mut self) -> Id {
+		self.updated = true;
```

Comment:
I was wondering if this needs to be updated only if there is any available id in the roaring bitmap?

Reply:
@arriqaaq Hey! Thanks for reviewing as always!!
As `updated` being a trigger to update the value containing the ID generator state, it needs to be updated also when `next_id` is incremented.
That said I think this code is correct- WDYT?